lib/src/main/java/com/exallium/rxrecyclerview/lib/RxRecyclerViewAdapter.java

```diff
@@ -132,20 +134,31 @@ public void onError(Throwable e) {
         @Override
         public void onNext(EventElement<K, V> rxEvent) {
             preProcessElement(rxEvent);
+            EventElement<K, V> currentRxEvent = null;
             switch (rxEvent.getData().getType()) {
                 case ADD:
-                    if (treeSet.add(rxEvent)) {
-                        notifyItemInserted(getIndexOf(rxEvent));
-                    } else {
-                        treeSet.remove(rxEvent);
+                    if((currentRxEvent = items.put(rxEvent.getData().getKey(), rxEvent)) != null) {
```

Comment:
Heh.  Should be a space between the if and ( as well... You can change it if you want otherwise I can handle it.